File: 19.Remove Nth Node From End of List.java
/*
 * @lc app=leetcode id=19 lang=java
 * @lcpr version=30005
 *
 * [19] Remove Nth Node From End of List
 */


// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // still not clear why is to use dummy head better and when should i use dummy head?
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode x = findFromBack(dummy, n + 1); // because i add dummy head, so it is actually (k+1)-th position from back, we count from dummy (-1)
        x.next = x.next.next; // why is .next -> .next.next not ...?

        return dummy.next; // why not head?
    }

    public ListNode findFromBack(ListNode head, int k) {
        ListNode p1 = head;
        for (int i = 0; i < k; i++) {
            p1 = p1.next;
        }
        ListNode p2 = head;
        while (p1 != null) {
            p1 = p1.next;
            p2 = p2.next;
        }
        return p2;
    }
}
// @lc code=end



/*
// @lcpr case=start
// [1,2,3,4,5]\n2\n
// @lcpr case=end

// @lcpr case=start
// [1]\n1\n
// @lcpr case=end

// @lcpr case=start
// [1,2]\n1\n
// @lcpr case=end

 */

File: 19.Remove Nth Node From End of List.py
#
# @lc app=leetcode id=19 lang=python3
# @lcpr version=30202
#
# [19] Remove Nth Node From End of List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # 1. dummy node
        dummy = ListNode(-1)
        dummy.next = head

        # 2. find the n-th node from end
        # x = findFromEnd(head, n)
        # you have to start from dummy node not head, why?
        # because we actually find the (n+1)-th node from end
        # so later we can easily remove n-th node
        # by x.next=x.next.next ==> we "jump" over the n-th node from end
        x = self.findFromEnd(dummy, n+1)

        # 3. remove the n-th node from end
        # we "jump" over the n-th node from end
        x.next = x.next.next

        # 4. return result
        return dummy.next

    def findFromEnd(self, head: ListNode, k: int) -> ListNode:
        # 1. get two pointers 
        p1 = head
        p2 = head

        # 2. let p1 move k steps
        for i in range(k):
            p1 = p1.next
        
        # 3. let p1 and p2 move together until p1 gets to the end/null
        while p1 != None:
            p1 = p1.next
            p2 = p2.next
        
        # 4. return p2 and it's at the exact position we want
        return p2


        
# @lc code=end



#
# @lcpr case=start
# [1,2,3,4,5]\n2\n
# @lcpr case=end

# @lcpr case=start
# [1]\n1\n
# @lcpr case=end

# @lcpr case=start
# [1,2]\n1\n
# @lcpr case=end

#

